keps/sig-instrumentation/20200415-cardinality-enforcement.md

```diff
@@ -0,0 +1,290 @@
+
+```prometheus
+# HELP some_metric alalalala
+# TYPE some_metric counter
+some_metric{label_too_many_values="1"} 1
+some_metric{label_too_many_values="2"} 1
+some_metric{label_too_many_values="3"} 1
+some_metric{label_too_many_values="4"} 1
+... //
+... // zillion metrics here
+some_metric{label_too_many_values="1000003"} 1
+```
+
+That would suck. With our cardinality enforcer in place, we would expect the output to look like this:
+
+```prometheus
+# HELP some_metric alalalala
+# TYPE some_metric counter
+some_metric{label_too_many_values="1"} 1
+some_metric{label_too_many_values="2"} 1
+some_metric{label_too_many_values="3"} 1
+some_metric{label_too_many_values="unexpected"} 1000000
+```
+
+We can effect this change by registering a whitelist to our prom registries during the boot-up sequence. Since we intercept metric registration events and have a wrapper around each of the primitive prometheus metric types, we can effectively add some code that looks like this (disclaimer: the below is pseudocode for demonstration):
+
+```golang
+const (
+	unexpectedLabel = "unexpected"
+)
+
+func (v *CounterVec) WithLabelValues(lvs ...string) CounterMetric {
+	if !v.IsCreated() {
+		return noop // return no-op counter
+	}
+	// INJECTED CODE HERE
+	newLabels = make([]string, len(lvs))
+	for i, l := range lvs {
+
+		// do we have a whitelist on any of the labels for this metric?
+		if metricLabelWhitelist, ok := MetricsLabelWhitelist[BuildFQName(v.CounterOpts.Namespace, v.CounterOpts.Subsystem, v.CounterOpts.Name)]; ok {
+			// do we have a whitelist for this label on this metric?
+			if whitelist, ok := metricLabelWhitelist[v.originalLabels[i]]; ok {
+				if whitelist.Has(l) {
+					newLabels[i] = l
+					continue
+				}
+			}
+		}
+		newLabels[i] = unexpectedLabel
+	}
+
+	return v.CounterVec.WithLabelValues(newLabels...)
+	// END INJECTED CODE
+}
+```
+
+This design allows us to optionally adopt @lilic's excellent idea about simplifying the interface for component owners, who can then opt to just specify a metric and label pair *without* having to specify a whitelist. Personally, I like that idea since it simplifies how a component owner can implement our cardinality enforcing helpers without having to necessary plumb through complicated maps. This would make it considerably easier to feed this data in through the command line since you could do something like this:
+
+```bash
+$ kube-apiserver --accepted-metric-labels "some_metric=label_too_many_values"
+```
+
+..which would then be interpreted by our machinery as this:
+
+
+```json
+[
+	{
+		"metric-id": "some_metric",
+		"label": "label_too_many_values",
+		"labelValueWhitelist": []
+	}
+]
+
+```
+
+## Open-Question
+_(Discussion Points which need to be resolved prior to merge)_
+
+- @dashpole
+
+> Should have labels with a specific set of values, should we start enforcing that all metrics have a whitelist at compile-time?
```

Comment:
This can be removed, or merged with @x13n's comment below.  As discussed above can't enforce that all labels have a whitelist, as some labels are determined at runtime (e.g. node name).